In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import re
import csv

In [2]:
ruta = ('C:/Users/Aaron/Desktop/Data_carga/data_spotify_clean.csv')
data_csv = pd.read_csv(ruta)
data_csv

track_id                 artists  \
0       5SuOikwiRyPMVoIQDJUgSV             Gen Hoshino   
1       4qPNDBW1i3p13qLCt0Ki3A            Ben Woodward   
2       1iJBSr7s7jYXzM8EGcbK5b  Ingrid Michaelson;ZAYN   
3       6lfxq3CG4xtTiEg7opyCyx            Kina Grannis   
4       5vjLSffimiIP26QG5WcN2K        Chord Overstreet   
...                        ...                     ...   
113995  2C3TZjDRiAzdyViavDJ217           Rainy Lullaby   
113996  1hIz5L4IB9hN3WRYPOCGPw           Rainy Lullaby   
113997  6x8ZfSoqDjuNa5SVP5QjvX           Cesária Evora   
113998  2e6sXL2bYv4bSz6VTdnfLs        Michael W. Smith   
113999  2hETkH7cOfqmz3LqZDHZf5           Cesária Evora   

                                               album_name  \
0                                                  Comedy   
1                                        Ghost (Acoustic)   
2                                          To Begin Again   
3       Crazy Rich Asians (Original Motion Picture Sou...   
4                                                 Hold On   
...                                                   ...   
113995  #mindfulness - Soft Rain for Mindful Meditatio...   
113996  #mindfulness - Soft Rain for Mindful Meditatio...   
113997                                            Best Of   
113998                                  Change Your World   
113999                                     Miss Perfumado   

                        track_name  popularity  duration_ms  explicit  \
0                           Comedy          73       230666     False   
1                 Ghost - Acoustic          55       149610     False   
2                   To Begin Again          57       210826     False   
3       Can't Help Falling In Love          71       201933     False   
4                          Hold On          82       198853     False   
...                            ...         ...          ...       ...   
113995         Sleep My Little Boy          21       384999     False   
113996            Water Into Light          22       385000     False   
113997              Miss Perfumado          22       271466     False   
113998                     Friends          41       283893     False   
113999                   Barbincor          22       241826     False   

        danceability  energy  key  loudness  mode  speechiness  acousticness  \
0              0.676  0.4610    1    -6.746     0       0.1430        0.0322   
1              0.420  0.1660    1   -17.235     1       0.0763        0.9240   
2              0.438  0.3590    0    -9.734     1       0.0557        0.2100   
3              0.266  0.0596    0   -18.515     1       0.0363        0.9050   
4              0.618  0.4430    2    -9.681     1       0.0526        0.4690   
...              ...     ...  ...       ...   ...          ...           ...   
113995         0.172  0.2350    5   -16.393     1       0.0422        0.6400   
113996         0.174  0.1170    0   -18.318     0       0.0401        0.9940   
113997         0.629  0.3290    0   -10.895     0       0.0420        0.8670   
113998         0.587  0.5060    7   -10.889     1       0.0297        0.3810   
113999         0.526  0.4870    1   -10.204     0       0.0725        0.6810   

        instrumentalness  liveness  valence    tempo  time_signature  \
0               0.000001    0.3580   0.7150   87.917               4   
1               0.000006    0.1010   0.2670   77.489               4   
2               0.000000    0.1170   0.1200   76.332               4   
3               0.000071    0.1320   0.1430  181.740               3   
4               0.000000    0.0829   0.1670  119.949               4   
...                  ...       ...      ...      ...             ...   
113995          0.928000    0.0863   0.0339  125.995               5   
113996          0.976000    0.1050   0.0350   85.239               4   
113997          0.000000    0.0839   0.7430  132.378               4   
113998          0.000000    0.2700

In [3]:
#TENGO QUE ESPECIFICAR EL ENCODING QUE SON DATOS UNICODE, QUIERE DECIR QUE ACEPTAN TODOS LOS CARACTERES
#Este es mi archivo de SQL EXPORTÁNDOLO PARA LIMPIAR
ruta2 = ('C:/Users/Aaron/Desktop/Data_carga/new_spotify.csv')
data_sql = pd.read_csv(ruta2,encoding='utf-16',delimiter='|')
data_sql

N_cod                track_id                      artists  \
0            1  5SuOikwiRyPMVoIQDJUgSV                  Gen Hoshino   
1            2  4qPNDBW1i3p13qLCt0Ki3A                 Ben Woodward   
2            3  1iJBSr7s7jYXzM8EGcbK5b       Ingrid Michaelson;ZAYN   
3            4  6lfxq3CG4xtTiEg7opyCyx                 Kina Grannis   
4            5  5vjLSffimiIP26QG5WcN2K             Chord Overstreet   
...        ...                     ...                          ...   
113995  113996  2tMjhETPLkWILXN3NnmDt3                 Chris Tomlin   
113996  113997  4VS6PK5bVEBtEt4lLvLYoi  Jesus Culture;Chris Quilala   
113997  113998  0s9o5zNpoRcYkXOVeLASCa        Hillsong Young & Free   
113998  113999  089YdobV7859anDvVu1TsY             Michael W. Smith   
113999  114000  6HKmAGpQgDNQ4sFhAgvkA2                    Starfield   

                                               album_name  \
0                                                  Comedy   
1                                        Ghost (Acoustic)   
2                                          To Begin Again   
3       Crazy Rich Asians (Original Motion Picture Sou...   
4                                                 Hold On   
...                                                   ...   
113995                                   Good Good Father   
113996                                 Let It Echo (Live)   
113997                             All Of My Best Friends   
113998                                 Decades of Worship   
113999                                          Starfield   

                        track_name  popularity  duration_ms  explicit  \
0                           Comedy          73       230666         0   
1                 Ghost - Acoustic          55       149610         0   
2                   To Begin Again          57       210826         0   
3       Can't Help Falling In Love          71       201933         0   
4                          Hold On          82       198853         0   
...                            ...         ...          ...       ...   
113995            Good Good Father          51       297146         0   
113996               Fierce - Live          43       246640         0   
113997       Best Friends - Studio          45       183414         0   
113998                 Awesome God          49       197613         0   
113999             Cry In My Heart          51       200880         0   

        danceability  energy  ...   loudness  mode  speechiness  acousticness  \
0              0.676  0.4610  ...  -6.746000     0       0.1430       0.03220   
1              0.420  0.1660  ... -17.235001     1       0.0763       0.92400   
2              0.438  0.3590  ...  -9.734000     1       0.0557       0.21000   
3              0.266  0.0596  ... -18.514999     1       0.0363       0.90500   
4              0.618  0.4430  ...  -9.681000     1       0.0526       0.46900   
...              ...     ...  ...        ...   ...          ...           ...   
113995         0.436  0.6000  ...  -4.400000     1       0.0273       0.29100   
113996         0.493  0.8140  ...  -6.366000     0       0.0439       0.00167   
113997         0.491  0.8280  ...  -5.368000     1       0.0311       0.08290   
113998         0.298  0.6600  ...  -8.187000     1       0.0410       0.28300   
113999         0.509  0.4120  ... -10.249000     1       0.0271       0.12300   

        instrumentalness  liveness  valence       tempo  time_signature  \
0               0.000001    0.3580    0.715   87.917000               4   
1               0.000006    0.1010    0.267   77.488998               4   
2               0.000000    0.1170    0.120   76.332001               4   
3               0.000071    0.1320    0.143  181.740005               3   
4               0.000000    0.0829    0.167  119.948997               4   
...                  ...       ...      ...         ...             ...   
113995          0.000000    0.0996    0.181  145

In [4]:
#TENEMOS LOS 2 ARCHIVOS data_sql y data_csv
#VAMOS A NORMALIZAR data_sql

In [5]:
#PODEMOS VER LAS VARIABLES CATEGORICAS ES DECIR DATOS TYPO OBJECT
data_sql.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114000 entries, 0 to 113999
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   N_cod             114000 non-null  int64  
 1   track_id          114000 non-null  object 
 2   artists           113999 non-null  object 
 3   album_name        113999 non-null  object 
 4   track_name        113999 non-null  object 
 5   popularity        114000 non-null  int64  
 6   duration_ms       114000 non-null  int64  
 7   explicit          114000 non-null  int64  
 8   danceability      114000 non-null  float64
 9   energy            113925 non-null  float64
 10  key               114000 non-null  int64  
 11  loudness          113910 non-null  float64
 12  mode              114000 non-null  int64  
 13  speechiness       114000 non-null  float64
 14  acousticness      109156 non-null  float64
 15  instrumentalness  114000 non-null  float64
 16  liveness          11

In [17]:
#Aquí elimino los registros faltantes debido a que no son completos, no es la manera más adecuada, debido a que podemos
#completarlos con modelos de machin learning para completar los datos(eso todavía no investigo), pero por motivos educativos los eliminaremos
data_sql.dropna(inplace=True)
data_sql.info()

<class 'pandas.core.frame.DataFrame'>
Index: 108871 entries, 0 to 113999
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   N_cod             108871 non-null  int64  
 1   track_id          108871 non-null  object 
 2   artists           108871 non-null  object 
 3   album_name        108871 non-null  object 
 4   track_name        108871 non-null  object 
 5   popularity        108871 non-null  int64  
 6   duration_ms       108871 non-null  int64  
 7   explicit          108871 non-null  int64  
 8   danceability      108871 non-null  float64
 9   energy            108871 non-null  float64
 10  key               108871 non-null  int64  
 11  loudness          108871 non-null  float64
 12  mode              108871 non-null  int64  
 13  speechiness       108871 non-null  float64
 14  acousticness      108871 non-null  float64
 15  instrumentalness  108871 non-null  float64
 16  liveness          108871 

Observamos que todas nuestras filas se quedaro con 108871 registros, después de la limpieza

In [20]:
#TOMO EN UNA LISTA LAS VARIABLES CATEGORICAS DE MI INTERÉS EN ESTE CASO, REALIZO UN FOR PARA RECORRER CADA COLUMNA
#data_sql[col] recorre todos los registros unicos de cada columna y me los devulve en un conteo
col_cat = ['artists','track_genre']
for col in col_cat:
    print(f'Columna {col} tiene {data_sql[col].nunique()} subniveles')

Columna artists tiene 30762 subniveles
Columna track_genre tiene 114 subniveles


In [21]:
#EL DESCRIBE ME MUESTRA DATOS ESTADÍSTICOS GENERALES DE LA DATA
data_sql.describe()

N_cod     popularity   duration_ms       explicit  \
count  108871.000000  108871.000000  1.088710e+05  108871.000000   
mean    57528.076687      33.525236  2.276981e+05       0.083558   
std     33004.918628      22.434675  1.061020e+05       0.276724   
min         1.000000       0.000000  8.586000e+03       0.000000   
25%     28827.500000      17.000000  1.740280e+05       0.000000   
50%     57938.000000      35.000000  2.124430e+05       0.000000   
75%     86180.500000      50.000000  2.608100e+05       0.000000   
max    114000.000000     100.000000  5.237295e+06       1.000000   

        danceability         energy            key       loudness  \
count  108871.000000  108871.000000  108871.000000  108871.000000   
mean        0.575318       0.629496       5.307033      -8.357791   
std         0.169087       0.248761       3.557740       5.040901   
min         0.000000       0.000000       0.000000     -49.530998   
25%         0.468000       0.461000       2.000000     -10.140000   
50%         0.588000       0.670000       5.000000      -7.110000   
75%         0.699000       0.837000       8.000000      -5.082000   
max         0.985000       1.000000      11.000000      -0.001000   

                mode    speechiness   acousticness  instrumentalness  \
count  108871.000000  108871.000000  108871.000000     108871.000000   
mean        0.638517       0.084420       0.328732          0.146235   
std         0.480432       0.107429       0.332908          0.302289   
min         0.000000       0.000000       0.000000          0.000000   
25%         0.000000       0.035600       0.025100          0.000000   
50%         1.000000       0.048100       0.192000          0.000028   
75%         1.000000       0.082400       0.616000          0.027900   
max         1.000000       0.965000       0.996000          1.000000   

            liveness        valence          tempo  time_signature  
count  108871.000000  108871.000000  108871.000000   108871.000000  
mean        0.211753       0.481421     121.898586        3.907147  
std         0.189735       0.258643      29.817617        0.419843  
min         0.000000       0.000000       0.000000        0.000000  
25%         0.097900       0.268000      98.815002        4.000000  
50%         0.130000       0.473000     121.994003        4.000000  
75%         0.269000       0.690000     140.035004        4.000000  
max         1.000000       0.995000     243.371994        5.000000

In [29]:
#CON LA FUNCIÓN DUPLICATED REVISO SI HAY FILAS DUPLICADAS Y LUEGO LO IMPRIMO EN UN VARIABLE
#No hay filas repetidas
filas_duplicados= data_sql.duplicated()
duplicados = data_sql[filas_duplicados]
duplicados

Empty DataFrame
Columns: [N_cod, track_id, artists, album_name, track_name, popularity, duration_ms, explicit, danceability, energy, key, loudness, mode, speechiness, acousticness, instrumentalness, liveness, valence, tempo, time_signature, track_genre]
Index: []

[0 rows x 21 columns]

In [34]:
#cantidad de mi data frame (FILAS Y COLUMNAS)
data_sql.shape

(108871, 21)

In [61]:
#Tenemos nombres de artistas que estan escritos así por ejemplo LISA,lisa,liSa, entonces vamos a colocarlos en formato Capitalize
artists = data_sql[data_sql['artists']=='LisA']
artists

Empty DataFrame
Columns: [N_cod, track_id, artists, album_name, track_name, popularity, duration_ms, explicit, danceability, energy, key, loudness, mode, speechiness, acousticness, instrumentalness, liveness, valence, tempo, time_signature, track_genre]
Index: []

[0 rows x 21 columns]

In [50]:
#AQUÍ SE MODIFICA LISA PASANDO A Lisa con lower y capitalize, solo para el campo artists y para LiSA
filas_lisa_index = data_sql[data_sql['artists']=='LiSA'].index
data_sql.loc[filas_lisa_index,'artists']= data_sql.loc[filas_lisa_index,'artists'].str.lower()
data_sql.loc[filas_lisa_index,'artists']= data_sql.loc[filas_lisa_index,'artists'].str.capitalize()
filas_lisa_index = data_sql[data_sql['artists']=='LiSA']
filas_lisa_index

Empty DataFrame
Columns: [N_cod, track_id, artists, album_name, track_name, popularity, duration_ms, explicit, danceability, energy, key, loudness, mode, speechiness, acousticness, instrumentalness, liveness, valence, tempo, time_signature, track_genre]
Index: []

[0 rows x 21 columns]

In [53]:
#HACEMOS LO MISMO PERO CON LISA
lisa = data_sql[data_sql['artists']=='LISA']
#lisa
filas_LISA_index = data_sql[data_sql['artists']=="LISA"].index
data_sql.loc[filas_LISA_index,'artists']=data_sql.loc[filas_LISA_index,'artists'].str.lower()
data_sql.loc[filas_LISA_index,'artists']=data_sql.loc[filas_LISA_index,'artists'].str.capitalize()

In [54]:
lisa2=data_sql[data_sql['artists']=='Lisa']
lisa2

N_cod                track_id artists                   album_name  \
9020    9021  08c9t8xZCZt7lVVTNeXpzn    Lisa                            炎   
11076  11077  60iYkEAcSpAFcsjRrgTlrc    Lisa                     明け星 / 白銀   
11856  11857  0essQdYYWaolOQq1MfVZ2o    Lisa                       NEW ME   
12352  12353  0f5hcKLCRgpdVV424X4XDQ    Lisa                           白銀   
12443  12444  2N1pcj1jkiRtDjpMwo0k4H    Lisa                     明け星 / 白銀   
14493  14494  4BvuZVf9KyBN3QiPfeI9hw    Lisa                    LANDSPACE   
14505  14506  0cSkn2l67csUljEy0EEBPn    Lisa                            炎   
14529  14530  0qMip0B2D4ePEjBJvAtYre    Lisa                     LEO-NiNE   
15098  15099  5Xs2wcsqZwZ6bfLSpAs08G    Lisa                          紅蓮華   
15192  15193  3SlQVRQAgsc6ac6UBM9dIk    Lisa                     LEO-NiNE   
15733  15734  3J34JtfZNylVXSKN4TDf12    Lisa                          明け星   
16010  16011  6RpBs7BsS9oiHKtzJXsMOS    Lisa  赤い罠(who loves it?) / ADAMAS   
16137  16138  0nUKU6e9COOCgO5tAUtlST    Lisa                          紅蓮華   
18679  18680  4cDm9C5jhBiIyYxVhQmMwB    Lisa                           往け   
19000  19001  7LqQO6uWQHDEue1ejeUVhF    Lisa                     Launcher   
19879  19880  45UI1yzA7vjkWHoAUZ8nGp    Lisa                 だってアタシのヒーロー。   
20155  20156  6vRkYTrWDzzVrZTqBJFR0u    Lisa          LiTTLE DEViL PARADE   
20925  20926  4WRoLfCYJ4sLMhXgQJnI2N    Lisa                 LOVER"S"MiLE   
57081  57082  4BvuZVf9KyBN3QiPfeI9hw    Lisa                    LANDSPACE   
57091  57092  0cSkn2l67csUljEy0EEBPn    Lisa                            炎   
57115  57116  0qMip0B2D4ePEjBJvAtYre    Lisa                     LEO-NiNE   
57615  57616  5Xs2wcsqZwZ6bfLSpAs08G    Lisa                          紅蓮華   
57645  57646  3J34JtfZNylVXSKN4TDf12    Lisa                          明け星   
57691  57692  3SlQVRQAgsc6ac6UBM9dIk    Lisa                     LEO-NiNE   
58217  58218  0nUKU6e9COOCgO5tAUtlST    Lisa                          紅蓮華   
58480  58481  6RpBs7BsS9oiHKtzJXsMOS    Lisa  赤い罠(who loves it?) / ADAMAS   
60406  60407  4cDm9C5jhBiIyYxVhQmMwB    Lisa                           往け   
60542  60543  7LqQO6uWQHDEue1ejeUVhF    Lisa                     Launcher   
61270  61271  45OX2jjEw1l7lOFJfDP9fv    Lisa                       LALISA   
61327  61328  2KZ3sNqPogEyMUUH6A5HFy    Lisa                       LALISA   
61640  61641  45UI1yzA7vjkWHoAUZ8nGp    Lisa                 だってアタシのヒーロー。   
62386  62387  6vRkYTrWDzzVrZTqBJFR0u    Lisa          LiTTLE DEViL PARADE   
63347  63348  4WRoLfCYJ4sLMhXgQJnI2N    Lisa                 LOVER"S"MiLE   
65356  65357  08c9t8xZCZt7lVVTNeXpzn    Lisa                            炎   
67956  67957  60iYkEAcSpAFcsjRrgTlrc    Lisa                     明け星 / 白銀   
77816  77817  45OX2jjEw1l7lOFJfDP9fv    Lisa                       LALISA   
92981  92982  0qMip0B2D4ePEjBJvAtYre    Lisa                     LEO-NiNE   

             track_name  popularity  duration_ms  explicit  danceability  \
9020                  炎          51       274412         0         0.477   
11076               明け星          50       269000         0         0.389   
11856            NEW ME          53       261238         0         0.274   
12352                白銀          51       264666         0         0.485   
12443                白銀          46       264666         0         0.485   
14493    crossing field          69       248946         0         0.450   
14505                 炎          67       275000         0         0.477   
14529               紅蓮華          74       237586         0         0.541   
15098     紅蓮華 -TV ver.-          60        89946         0         0.559   
15192         unlasting          59       296813         0         0.637   
15733               明け星          61       269000         0         0.389   
16010            ADAMAS          63       225000         0         0.514   
16137               紅蓮華          56       238586         0         0.544   
18679           

lo mismo tenemos que hacer con los datos de 
1	Desejo de Menina
2	Deadlife
3	DJ Fixx;Ondamike
4	LOS HERMANOS
5	Surade|3AM VIBES
6	Ralphi Rosario;Donna Blakely;lego
7	TERRA
8	Shine
9	NaN
10	FISHER

In [56]:
#HACEMOS LO MISMO PARA DEADLIFE
filas_index = data_sql[data_sql['artists']=="DEADLIFE"].index
data_sql.loc[filas_index,'artists']=data_sql.loc[filas_index,'artists'].str.lower()
data_sql.loc[filas_index,'artists']=data_sql.loc[filas_index,'artists'].str.capitalize()

In [82]:
# otro método usando replace
data_sql.loc[data_sql['artists'].isin(['Desejo de menina']),'artists']='Desejo De Menina'

In [89]:
#FILTRO PARA MOSTRAR LOS DATOS POR EL NOMBRE DEL ARTISTA
f1=data_sql[data_sql['artists']=='Shine']
f1

N_cod                track_id artists         album_name  \
19187    19188  2SLwPaJTN6B6zwcIwsuQAK   Shine  The Best Of Shine   
112255  112256  5z0o7Ep4MktEA00g804pC3   Shine     Street Of Gold   

            track_name  popularity  duration_ms  explicit  danceability  \
19187              燕尾蝶          46       228530         0         0.548   
112255  Street Of Gold          30       183396         0         0.883   

        energy  ...  loudness  mode  speechiness  acousticness  \
19187    0.491  ...    -5.761     1       0.0236         0.596   
112255   0.544  ...   -13.627     0       0.4120         0.804   

        instrumentalness  liveness  valence       tempo  time_signature  \
19187              0.000     0.144    0.417   82.070999               4   
112255             0.806     0.109    0.532  105.940002               4   

        track_genre  
19187      cantopop  
112255        study  

[2 rows x 21 columns]

In [94]:
#GUARDO MI NUEVO ARCHIVO CSV PARA IMPORTAR A MI BASE DE DATOS
data_sql.to_csv('data_new_spotify.csv',index=False)